pkg/webhook/clusterset/validatingWebhook.go

```diff
@@ -0,0 +1,356 @@
+// function
+func (a *AdmissionHandler) serve(w io.Writer, r *http.Request, admit admitFunc) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		klog.Errorf("contentType=%s, expect application/json", contentType)
+		return
+	}
+
+	klog.V(2).Info(fmt.Sprintf("handling request: %s", body))
+
+	// The AdmissionReview that was sent to the webhook
+	requestedAdmissionReview := v1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned
+	responseAdmissionReview := v1.AdmissionReview{}
+
+	deserializer := options.Codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		klog.Error(err)
+		responseAdmissionReview.Response = toAdmissionResponse(err)
+	} else {
+		// pass to admitFunc
+		responseAdmissionReview.Response = admit(requestedAdmissionReview.Request)
+	}
+
+	responseAdmissionReview.Kind = requestedAdmissionReview.Kind
+	responseAdmissionReview.APIVersion = requestedAdmissionReview.APIVersion
+	// Return the same UID
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+
+	klog.V(2).Info(fmt.Sprintf("sending response: %+v", responseAdmissionReview))
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		klog.Error(err)
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		klog.Error(err)
+	}
+}
+
+var managedClustersGVR = metav1.GroupVersionResource{
+	Group:    "cluster.open-cluster-management.io",
+	Version:  "v1",
+	Resource: "managedclusters",
+}
+var clusterDeploymentsGVR = metav1.GroupVersionResource{
+	Group:    "hive.openshift.io",
+	Version:  "v1",
+	Resource: "clusterdeployments",
+}
+var clusterPoolsGVR = metav1.GroupVersionResource{
+	Group:    "hive.openshift.io",
+	Version:  "v1",
+	Resource: "clusterpools",
+}
+var clusterClaimsGVR = metav1.GroupVersionResource{
+	Group:    "hive.openshift.io",
+	Version:  "v1",
+	Resource: "clusterclaims",
+}
+
+const (
+	clusterSetLabel = "cluster.open-cluster-management.io/clusterset"
+)
+
+func (a *AdmissionHandler) validateResource(request *v1.AdmissionRequest) *v1.AdmissionResponse {
+	status := &v1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	switch request.Resource {
+	case managedClustersGVR:
+	case clusterDeploymentsGVR:
+	case clusterClaimsGVR:
+	case clusterPoolsGVR:
+	default:
+		return status
+	}
+
+	switch request.Operation {
+	case v1.Create:
+		allowAll, err := a.allowWithoutClusterSetLabel(request.UserInfo)
+		if allowAll {
+			return status
+		} else if err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+				Message: err.Error(),
+			}
+			return status
+		}
+		return a.validateCreateRequest(request)
+	case v1.Update:
```

Comment:
this also need a.allowWithoutClusterSetLabel(request.UserInfo)

Reply:
done. move a.allowWithoutClusterSetLabel(request.UserInfo) before Create and Update.